src/Mod/Sketcher/Gui/SnapManager.cpp

```diff
@@ -0,0 +1,309 @@
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+#include <QApplication>
+#endif  // #ifndef _PreComp_
+
+#include <Mod/Sketcher/App/SketchObject.h>
+
+#include "SnapManager.h"
+#include "ViewProviderSketch.h"
+
+
+using namespace SketcherGui;
+using namespace Sketcher;
+
+/************************************ Attorney *******************************************/
+
+inline int ViewProviderSketchSnapAttorney::getPreselectPoint(const ViewProviderSketch& vp)
+{
+    return vp.getPreselectPoint();
+}
+
+inline int ViewProviderSketchSnapAttorney::getPreselectCross(const ViewProviderSketch& vp)
+{
+    return vp.getPreselectCross();
+}
+
+inline int ViewProviderSketchSnapAttorney::getPreselectCurve(const ViewProviderSketch& vp)
+{
+    return vp.getPreselectCurve();
+}
+
+/**************************** ParameterObserver nested class *****************************/
+SnapManager::ParameterObserver::ParameterObserver(SnapManager& client) : Client(client)
+{
+    initParameters();
+    subscribeToParameters();
+}
+
+SnapManager::ParameterObserver::~ParameterObserver()
+{
+    unsubscribeToParameters();
+}
+
+void SnapManager::ParameterObserver::initParameters()
+{
+    // static map to avoid substantial if/else branching
+    //
+    // key->first               => String of parameter,
+    // key->second              => Update function to be called for the parameter,
+    str2updatefunction = {
+        {"SnapToObjects",
+            [this](const std::string& param) {updateSnapToObjectParameter(param); }},
+        {"SnapAngle",
+            [this](const std::string& param) {updateSnapAngleParameters(param); }},
+    };
+
+    for (auto& val : str2updatefunction) {
+        auto string = val.first;
+        auto function = val.second;
+
+        function(string);
+    }
+}
+
+void SnapManager::ParameterObserver::updateSnapToObjectParameter(const std::string& parametername)
+{
+    (void)parametername;
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+
+    Client.snapToObjects = hGrp->GetBool("SnapToObjects", true);
+}
+
+void SnapManager::ParameterObserver::updateSnapAngleParameters(const std::string& parametername)
+{
+    (void)parametername;
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+
+    Client.snapAngle = hGrp->GetFloat("SnapAngle", 5.) * M_PI / 180;
+}
+
+
+void SnapManager::ParameterObserver::subscribeToParameters()
+{
+    try {
+        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+        hGrp->Attach(this);
+    }
+    catch (const Base::ValueError& e) { // ensure that if parameter strings are not well-formed, the exception is not propagated
+        Base::Console().Error("SnapManager: Malformed parameter string: %s\n", e.what());
+    }

```

Comment:
I'm not sure what this is catching. Was it copied from somewhere ?

Reply:
Probably from CoinManager.

GetParameterGroupByPath will raise an exception if the string is malformed.

Reply:
Yes it was like this in CoinManager. 
Not sure if this is really useful as the strings are not likely to be malformed as they are hardcoded, it's not a user input. Should I leave them or remove them?

Reply:
When I wrote CoinManager, I had in mind to reuse that code to create a generic parameter observer using the std::map pattern. Such class would be given a path and the path can be malformed.

Hardcoded strings can also be malformed, or they can become malformed if the parameter class changes (but then there will be bigger problems).

The biggest problem if a string is malformed and the exception is not caught, is that the lines of code just after subscribe to Parameter(), all the way back to core where the exception will be ultimately caught and an error message presented, will not be executed.

If the exception is caught here, those lines will be executed. The parameters will simply not update, because the observer could not attach.

So, I do agree that the risk is very small. However, because the functions used can generate an exception, the function cannot be declared noexcept either to improve performance. So I do not quite see a reason why to remove it. Specially when it is encapsulated. 

But it is indeed not the most important aspect of the PR.